### PR TITLE
[SYCL-MLIR]: Use alias analysis in WhileLICM transformation 

### PR DESCRIPTION
**Pull request by etiotto**
Currently the polygeist `WhileLICM` transformation does not make use of the MLIR alias analysis and instead uses its own (simple) analysis to determine whether an operation in the loop is loop invariant. 

This PR enhances `WhileLICM` to make use of the SYCL  `AliasAnalysis`. With this change, a load that uses as base pointer an argument with the `llvm.noalias` attribute, can be hoisted out of the `scf.while` loop, allowing it to be later transformed into a `scf.for` loop

### CODE REVIEW

**Comment by etiotto** [comment]
Note: depends on https://github.com/intel/llvm/pull/8863

File: polygeist/test/polygeist-opt/canonicalizefor.mlir
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/canonicalizefor.mlir:87]: why remove the attributes?

File: polygeist/test/polygeist-opt/canonicalizefor.mlir
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/canonicalizefor.mlir:192]: please align the operations in this function.

**Review by whitneywhtsang — COMMENTED** [review]
Not yet review CanonicalizeFor.cpp.

**Review by whitneywhtsang — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
The L0 GEN9 LLVM Test Suite failure is not real, as `image_max_size.cpp` is flaky, it sometimes time out, and sometimes not.
Created https://github.com/intel/llvm/pull/8899 to mark that test case as unsupported on L0.

**Comment by etiotto** [comment]
> The L0 GEN9 LLVM Test Suite failure is not real, as `image_max_size.cpp` is flaky, it sometimes time out, and sometimes not. Created #8899 to mark that test case as unsupported on L0.

#8899 has landed. So we can merge this one once we get another approval.

**Review by victor-eds — APPROVED** [review]
LGTM. However, his makes me wonder whether we can avoid some duplication between passes working on loops.

**Comment by etiotto** [comment]
> LGTM. However, his makes me wonder whether we can avoid some duplication between passes working on loops and hoisting values.

Yep. We have our LICM pass as well as the WhileLICM pass (that existed in MLIR already). I tactically chose to enhance the later in this PR but I also think we should have one way to do LICM. 

We could enhance the LICM pass to make it work on `scf.while` loops and then get rid of `WhileLICM`.

**Pull request merged**
